AX: DidSpellCheck attribute not applied for editable pages/documents

hoffmanjoshua opened the pull request:
#### 08cc264f5e2337bed167ab1d39c54057540c5bc2
<pre>
AX: DidSpellCheck attribute not applied for editable pages/documents
<a href="https://bugs.webkit.org/show_bug.cgi?id=294103">https://bugs.webkit.org/show_bug.cgi?id=294103</a>
<a href="https://rdar.apple.com/149710398">rdar://149710398</a>

Reviewed by Tyler Wilcock.

We did not previously consider an editable web area (page) to be an editible
ancestor, so the AXDidSpellCheck attribute would not be applied. This caused
certain ATs to not read out misspellings, if they handle the spellchecking.

This PR adds a new check to editableAncestor to return true if we hit a
web area that is also editable. New testing infrastructure was also added to
allow this attribute to be visible to tests that use the
attributedStringForTextMarkerRangeWithDidSpellCheck method.

* LayoutTests/accessibility/mac/designmode-lazy-spellchecking-expected.txt: Added.
* LayoutTests/accessibility/mac/designmode-lazy-spellchecking.html: Added.
* LayoutTests/accessibility/mac/editable-page-lazy-spellchecking-expected.txt: Added.
* LayoutTests/accessibility/mac/editable-page-lazy-spellchecking.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::Accessibility::editableAncestor):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handlePageEditibilityChanged):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::isEditableWebArea const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setDesignMode):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setEditable):
* Source/WebCore/page/Page.h:
(WebCore::Page::setEditable): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageSetEditable):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TextInputController.idl:
* Tools/WebKitTestRunner/InjectedBundle/TextInputController.cpp:
(WTR::TextInputController::setPageEditable):
* Tools/WebKitTestRunner/InjectedBundle/TextInputController.h:
* Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityUIElementAtspi.cpp:
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::createJSStringRef):
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRange):
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithOptions):
* Tools/WebKitTestRunner/InjectedBundle/playstation/AccessibilityUIElementPlayStation.cpp:
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):
* Tools/WebKitTestRunner/InjectedBundle/win/AccessibilityUIElementWin.cpp:
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):

Canonical link: <a href="https://commits.webkit.org/295998@main">https://commits.webkit.org/295998@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b60dee6c7d9b27f71bfb57fa8e8a3552745dcf87

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26435 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16832 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111894 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57282 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108723 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27105 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34937 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81013 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109688 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21505 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96270 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61352 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20951 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14377 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56736 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14406 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114792 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33822 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24936 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90075 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34186 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92501 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89784 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/22947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34711 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12521 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29449 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33747 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39160 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33493 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36846 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35092 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
```diff
@@ -649,6 +651,8 @@ class AXIsolatedTree : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<AX
     // Set to true by the AXObjectCache and false by AXIsolatedTree.
     bool m_relationsNeedUpdate { true };
 
+    std::atomic<bool> m_pageIsEditable { false };
```
As-is, I don't think this needs to be atomic. The only time this is touched by the main-thread is before the tree is stored via `storeTree`, and the accessibility thread cannot possibly use an AXIsolatedTree before `storeTree`.

However, I'm wondering if we need to update this bool after dynamic page changes? Isn't it possible for embedders of WebKit to set the editability state of the page to false, to true, to false?

minorninth replied:
As far as I know, the page editability could change at any time. From JS you can add contenteditable=true on the body or html element, or you can call document.designMode = "on". And yes, there are also APIs to do this when embedding WebKit that might be subtly different but to accessibility I think they're the same.

I think we should treat this as just another attribute and use our existing mechanisms to sync it to the isolated tree.

I think it'd be simplest to just make it a bool attribute in the property map, just one that happens to only apply to a web area.

In the future if we wanted, we could make a new property map for the whole tree.

hoffmanjoshua replied:
Moved to storing this as a property on AXIsolatedObject (only for web areas), after offline discussion.

This property also takes into account the `designmode` status of the document.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -383,6 +383,15 @@ bool AccessibilityObject::isARIATextControl() const
     return ariaRoleAttribute() == AccessibilityRole::TextArea || ariaRoleAttribute() == AccessibilityRole::TextField || ariaRoleAttribute() == AccessibilityRole::SearchField;
 }
 
+bool AccessibilityObject::isWebAreaAndEditable() const
```
Would `isEditableWebArea` convey the same meaning in a more succinct way?

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -177,6 +177,10 @@ RefPtr<AXIsolatedTree> AXIsolatedTree::create(AXObjectCache& axObjectCache)
     }
 
+    auto* page = document->page();
+    if (page)
```
if (auto* page = document->page())

hoffmanjoshua committed:
AX: DidSpellCheck attribute not applied for editable pages/documents
https://bugs.webkit.org/show_bug.cgi?id=294103
rdar://149710398

Reviewed by Tyler Wilcock.

We did not previously consider an editable web area (page) to be an editible
ancestor, so the AXDidSpellCheck attribute would not be applied. This caused
certain ATs to not read out misspellings, if they handle the spellchecking.

This PR adds a new check to editableAncestor to return true if we hit a
web area that is also editable. New testing infrastructure was also added to
allow this attribute to be visible to tests that use the
attributedStringForTextMarkerRangeWithDidSpellCheck method.

* LayoutTests/accessibility/mac/designmode-lazy-spellchecking-expected.txt: Added.
* LayoutTests/accessibility/mac/designmode-lazy-spellchecking.html: Added.
* LayoutTests/accessibility/mac/editable-page-lazy-spellchecking-expected.txt: Added.
* LayoutTests/accessibility/mac/editable-page-lazy-spellchecking.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::Accessibility::editableAncestor):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handlePageEditibilityChanged):
(WebCore::AXObjectCache::updateIsolatedTree):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::isEditableWebArea const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setDesignMode):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setEditable):
* Source/WebCore/page/Page.h:
(WebCore::Page::setEditable): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageSetEditable):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TextInputController.idl:
* Tools/WebKitTestRunner/InjectedBundle/TextInputController.cpp:
(WTR::TextInputController::setPageEditable):
* Tools/WebKitTestRunner/InjectedBundle/TextInputController.h:
* Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityUIElementAtspi.cpp:
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityUIElementIOS.mm:
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::createJSStringRef):
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRange):
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithOptions):
* Tools/WebKitTestRunner/InjectedBundle/playstation/AccessibilityUIElementPlayStation.cpp:
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):
* Tools/WebKitTestRunner/InjectedBundle/win/AccessibilityUIElementWin.cpp:
(WTR::AccessibilityUIElement::attributedStringForTextMarkerRangeWithDidSpellCheck):

Canonical link: https://commits.webkit.org/295998@main
